packages/application/src/shell.ts

```diff
@@ -289,4 +508,262 @@ namespace Private {
+      const index = this._findInsertIndex(item);
+      ArrayExt.insert(this._items, index, item);
+      this._widgetPanel.insertWidget(index, widget);
+      this.updateMenu();
+      this._refreshVisibility();
+
+      if (this._addPaletteEntry) {
+        this._addPaletteEntry(widget, this._area);
+      }
+    }
+
+    /**
+     * Hide the side panel
+     */
+    hide(): void {
+      this._isHiddenByUser = true;
+      this._refreshVisibility();
+    }
+
+    /**
+     * Show the side panel
+     */
+    show(): void {
+      this._isHiddenByUser = false;
+      this._refreshVisibility();
+    }
+
+    /**
+     * Update menu entries
+     */
+    updateMenu(): void {
+      if (!this._updateMenu) {
+        return;
+      }
+      if (this._disposableMenu) {
+        this._disposableMenu.dispose();
+      }
+      this._disposableMenu = this._updateMenu(this._area, this._menuEntryLabel);
+    }
+
+    /**
+     * Find the insertion index for a rank item.
+     */
+    private _findInsertIndex(item: Private.IRankItem): number {
+      return ArrayExt.upperBound(this._items, item, Private.itemCmp);
+    }
+
+    /**
+     * Find the index of the item with the given widget, or `-1`.
+     */
+    private _findWidgetIndex(widget: Widget): number {
+      return ArrayExt.findFirstIndex(this._items, i => i.widget === widget);
+    }
+
+    /**
+     * Find the widget with the given id, or `null`.
+     */
+    private _findWidgetByID(id: string): Widget | null {
+      const item = find(this._items, value => value.widget.id === id);
+      return item ? item.widget : null;
+    }
+
+    /**
+     * Refresh the visibility of the stacked panel.
+     */
+    private _refreshVisibility(): void {
+      this._panel.setHidden(this._isHiddenByUser);
+      this._updated.emit();
+    }
+
+    /*
+     * Handle the `widgetRemoved` signal from the panel.
+     */
+    private _onWidgetRemoved(sender: StackedPanel, widget: Widget): void {
+      if (widget === this._lastCurrent) {
+        this._lastCurrent = null;
+      }
+      ArrayExt.removeAt(this._items, this._findWidgetIndex(widget));
+
+      this.updateMenu();
+      this._refreshVisibility();
+
+      if (this._removePaletteEntry) {
+        this._removePaletteEntry(widget, this._area);
+      }
+    }
+
+    private _area: 'left' | 'right';
+    private _isHiddenByUser = false;
+    private _items = new Array<Private.IRankItem>();
+    private _panel: Panel;
+    private _widgetPanel: StackedPanel;
+    private _current: Widget | null;
+    private _lastCurrent: Widget | null;
+    private _updated: Signal<SideBarHandler, void> = new Signal(this);
+    private _menuEntryLabel: string;
+    private _disposableMenu: IDisposable | null = null;
+    private _updateMenu: SideBarPanel.UpdateSideBarMenuFn | null = null;
+    private _addPaletteEntry: SideBarPanel.AddPaletteEntryFn | null = null;
+    private _removePaletteEntry: SideBarPanel.RemovePaletteEntryFn | null = null;
```

Comment:
Wondering if there is a way to make these more generic and not specific to the menu and command palette only?

It looks like the plugin code (in `@jupyter-notebook/application-extension`) interacting with the menu and command palette need to know when a new widget is added or removed from the shell? So maybe exposing such generic signals might be enough?

Reply:
Agree, thanks for the suggestion.
https://github.com/jupyter/notebook/pull/6487/commits/dd46b05259d0894dc81498511f44dc62b53cf80e includes the use of signals.

Reply:
Nice thanks I'll check it out but this is looking good.